src/signTxCatalystRegistration.c

```diff
@@ -0,0 +1,576 @@
+		}
+
+		signTxCatalystRegistration_handleVotingRewardsAddress_addressParams_ui_runStep();
+	}
+}
+
+// ============================== NONCE ==============================
+
+enum {
+	HANDLE_NONCE_STEP_DISPLAY = 8600,
+	HANDLE_NONCE_STEP_RESPOND,
+	HANDLE_NONCE_STEP_INVALID,
+};
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleNonce_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTxCatalystRegistration_handleNonce_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_NONCE_STEP_DISPLAY) {
+		STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+
+		ui_displayUint64Screen(
+		        "Nonce",
+		        subctx->stateData.nonce,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_NONCE_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_NONCE_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleNonceAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// sanity checks
+		CHECK_STATE(STATE_CATALYST_REGISTRATION_NONCE);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		explicit_bzero(&subctx->stateData, SIZEOF(subctx->stateData));
+	}
+	{
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		VALIDATE(wireDataSize == 8, ERR_INVALID_DATA);
+		subctx->stateData.nonce = u8be_read(wireDataBuffer);
+		TRACE(
+		        "Catalyst registration nonce: %d",
+		        subctx->stateData.nonce
+		);
+	}
+	{
+		auxDataHashBuilder_catalystRegistration_addNonce(auxDataHashBuilder, subctx->stateData.nonce);
+	}
+
+	// TODO - is it worth declaring a policy for this?
+	subctx->ui_step = HANDLE_NONCE_STEP_DISPLAY;
+	signTxCatalystRegistration_handleNonce_ui_runStep();
+}
+
+// ============================== CONFIRM ==============================
+
+enum {
+	HANDLE_CONFIRM_STEP_FINAL_CONFIRM,
+	HANDLE_CONFIRM_STEP_DISPLAY_HASH,
+	HANDLE_CONFIRM_STEP_RESPOND,
+	HANDLE_CONFIRM_STEP_INVALID,
+};
+
+static void signTxCatalystRegistration_handleConfirm_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTxCatalystRegistration_handleConfirm_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+	UI_STEP(HANDLE_CONFIRM_STEP_FINAL_CONFIRM) {
+		ui_displayPrompt(
+		        "Confirm voting key",
+		        "registration?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_CONFIRM_STEP_DISPLAY_HASH) {
+		ui_displayHexBufferScreen(
+		        "Auxiliary data hash",
+		        (uint8_t*) &ctx->auxDataHash,
+		        SIZEOF(ctx->auxDataHash),
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_CONFIRM_STEP_RESPOND) {
+		struct {
+			uint8_t auxDataHash[AUX_DATA_HASH_LENGTH];
+			uint8_t signature[ED25519_SIGNATURE_LENGTH];
+		} wireResponse;
+
+		ASSERT(SIZEOF(ctx->auxDataHash) == AUX_DATA_HASH_LENGTH);
```

Comment:
What about asserting `SIZEOF(ctx->auxDataHash)` too? Also couple of lines below. And the asserts can be made static. 
(Might be worth checking if they are static across all the codebase.)

Reply:
This line above is doing just that - did you have something else in mind? I changed it to a static assert nevertheless

Reply:
I guess I meant the other variable (source/destination of os_memmove), it it not relevat anymore with view_memmove.